### ISSUE TITLE
Document use of kairos-agent upgrade for dir and file option

### ISSUE THREAD

**Issue by swetasgit** [kairos-io/kairos#2158]
**Kairos version:**

PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
KAIROS_FLAVOR="debian"
KAIROS_REGISTRY_AND_ORG="quay.io/kairos"
KAIROS_ID_LIKE="kairos-standard-debian-bookworm"
KAIROS_ARTIFACT="kairos-debian-bookworm-standard-amd64-generic-v2.5.0-k3sv1.27.9+k3s1"
KAIROS_HOME_URL="https://github.com/kairos-io/kairos"
KAIROS_SOFTWARE_VERSION="v1.27.9+k3s1"
KAIROS_ID="kairos"
KAIROS_NAME="kairos-standard-debian-bookworm"
KAIROS_IMAGE_REPO="quay.io/kairos/debian:bookworm-standard-amd64-generic-v2.5.0-k3sv1.27.9-k3s1"
KAIROS_IMAGE_LABEL="bookworm-standard-amd64-generic-v2.5.0-k3sv1.27.9-k3s1"
KAIROS_FLAVOR_RELEASE="bookworm"
KAIROS_VARIANT="standard"
KAIROS_GITHUB_REPO="kairos-io/kairos"
KAIROS_VERSION="v2.5.0-v1.27.9-k3s1"
KAIROS_PRETTY_NAME="kairos-standard-debian-bookworm v2.5.0-v1.27.9-k3s1"
KAIROS_TARGETARCH="amd64"
KAIROS_RELEASE="v2.5.0"
KAIROS_BUG_REPORT_URL="https://github.com/kairos-io/kairos/issues"
KAIROS_SOFTWARE_VERSION_PREFIX="k3s"
KAIROS_VERSION_ID="v2.5.0-v1.27.9-k3s1"
KAIROS_MODEL="generic"

**CPU architecture, OS, and Version:**

Linux localhost 6.1.0-17-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30) x86_64 GNU/Linux
**Describe the bug**

not able to upgrade using command:
kairos-agent --debug upgrade file:<Image.tar>
Giving error : 
DEBU[2024-01-20T13:54:03Z] Running cmd: 'tune2fs -L COS_ACTIVE /run/initramfs/cos-state/cOS/transition.img'
ERRO[2024-01-20T13:54:04Z] Failed to apply label COS_ACTIVE to $s%!(EXTRA string=/run/initramfs/cos-state/cOS/transition.img)
ERRO[2024-01-20T13:54:05Z] Failed deploying image to file '/run/initramfs/cos-state/cOS/transition.img': exit status 1
DEBU[2024-01-20T13:54:05Z] Unmounting partition COS_RECOVERY
DEBU[2024-01-20T13:54:05Z] Mounting partition COS_STATE
1 error occurred:
        * exit status 1
DIR ERROR: 
sudo kairos-agent --debug upgrade dir:/home/kairos/upgrade/
getting successfully upgraded as per log message but not able to boot up after restart /boot/vmlinuz not found  
**To Reproduce**

docker save -o <image.tar> <latest image from repo>
copy image.tar to kairos
execute : sudo kairos-agent --debug upgrade file:IMAGE.tar

For DIR
copy image.tar to kairos system
create directory and untar the image
execute : sudo kairos-agent --debug upgrade dir:/home/kairos/upgrade/
reboot the system.  
**Expected behavior**
System is upgraded

**Logs**

<img width="386" alt="dir-error" src="https://github.com/kairos-io/kairos/assets/83767840/268a0726-514c-4492-aad7-996104660ca9">
[filelogs.txt](https://github.com/kairos-io/kairos/files/13998185/filelogs.txt)
[dirlogs.txt](https://github.com/kairos-io/kairos/files/13998186/dirlogs.txt)

**Additional

**Comment by Itxaka** [comment]
This is expected, the file upgrade is not for docker artifacts and the dir is also not for docker artifacts. 

To upgrade from a dir, you need to use `kairos-agent pull-image` to pull the full image, otherwise the docker.tar file has layers and metadata on it that dont really work with our upgrade.



File upgrade is to upgrade with a provided image like the active.img (so an extt2 formatted image) or a squashfs image, not for a docker extracted image (docker.tar)

There is a feature request to provide a different source type which allows for this type of extracted tar (docker save format) but no work has been done on it.

**Comment by swetasgit** [comment]
@Itxaka  thanks for sharing the information. Our requirement is that we are customizing reusing kairos base image to recreate custom iso with auroraboot and so we want to upgrade the kairos agent offline using the docker image tar. I hope we get this feature in the upcoming release. 
Meanwhile to upgrade offline if I want to pull image via kairos-agent pull-image i get error:
root@localhost:~/upgrade# kairos-agent pull-image quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1 -l
INFO[2024-01-30T09:33:15Z] kairos-agent version v2.7.5
INFO[2024-01-30T09:33:15Z] Starting download and extraction for image quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1 to /root/upgrade/-l
GET https://quay.io/v2/kairos/debian/manifests/bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1: MANIFEST_UNKNOWN: manifest unknown; map[]

Am I pulling the right image from the right location? Also, how can I create a custom image from our customized kairos docker image which can be used for upgrade?

**Comment by Itxaka** [comment]
> @Itxaka thanks for sharing the information. Our requirement is that we are customizing reusing kairos base image to recreate custom iso with auroraboot and so we want to upgrade the kairos agent offline using the docker image tar. I hope we get this feature in the upcoming release. Meanwhile to upgrade offline if I want to pull image via kairos-agent pull-image i get error: root@localhost:~/upgrade# kairos-agent pull-image quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1 -l INFO[2024-01-30T09:33:15Z] kairos-agent version v2.7.5 INFO[2024-01-30T09:33:15Z] Starting download and extraction for image quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1 to /root/upgrade/-l GET [quay.io/v2/kairos/debian/manifests/bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1](https://quay.io/v2/kairos/debian/manifests/bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1): MANIFEST_UNKNOWN: manifest unknown; map[]
> 
> Am I pulling the right image from the right location? Also, how can I create a custom image from our customized kairos docker image which can be used for upgrade?

no, you are missing the kubernetes version in the image, see all the images available at https://quay.io/repository/kairos/debian?tab=tags

it should be one of:
 - quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1.29.0-k3s1
 - quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1.27.9-k3s1
 - quay.io/kairos/debian:bookworm-standard-amd64-generic-v3.0.0-alpha1-k3sv1.28.5-k3s1

To create a customized image you just have to use one of those as your base image, same as you build your own image and read the docs about the customization: https://kairos.io/docs/advanced/customizing/

But if you are already building your own image for install it should be the same.

**Comment by swetasgit** [comment]
Thanks  @Itxaka . The dir option works for me and also moving the tar to a remote location and upgrading. I am closing this issue. But could you please update the documentation with the details shared.

**Comment by Itxaka** [comment]
> Thanks @Itxaka . The dir option works for me and also moving the tar to a remote location and upgrading. I am closing this issue. But could you please update the documentation with the details shared.

will do. Pretty sure there is a card somewhere for this but cant finds it, so maybe we should keep this open and tag i documentation so we get to it

**Comment by cheese** [comment]
related to #1509

**Comment by mudler-agent** [comment]
This is still relevant, but the scope has changed since the original report. Current `kairos-agent` accepts `oci:`, `ocifile:`, `dir:`, and `file:` upgrade sources, while the manual upgrade page only shows the generic `<type>:<address>` form and an OCI example. It does not explain the expected artifact for each local source, which caused the original confusion between a Docker archive, an extracted root filesystem, and a filesystem image. I suggest moving this documentation task to `kairos-io/kairos-docs` and updating `docs/upgrade/manual.md`, including an offline `ocifile:` example. This issue is not superseded yet.
